### PR TITLE
Fix/gff outputs

### PR DESCRIPTION
**Pull request by Ales-ibt**
Final changes before merge and release new version!!!

  Combined report (*_combined_report.tsv)
  - Added a contig_id column (after protein_id).
  - Added a summary_string column condensing each gene's annotations into comma-joined tags vf,arg,mge,bgc.

  Derived GFFs
  - Fixed *_user_mobilome_clean.gff: it was missing a header and contained every gene; now it has ##gff-version 3 and only genes covered by an MGE.
  - Fixed *_user_mobilome_extra.gff header to just ##gff-version 3 (the ##sequence-region lines stay only in full).
  - Inject the report's summary_string into the derived GFFs as a pathofact2=<summary> attribute on matching CDSs.
  - Produce the clean/extra/full GFFs even when no user proteins are supplied, built from the Prodigal/tRNA genes GFF and named *_mobilome_{clean,extra,full}.gff.gz

BTW, this PR fix issue #95 

Thanks for reviewing!

### CODE REVIEW

File: bin/gff_mapping.py [hunk 1]
```diff
@@ -254,10 +324,10 @@ def gff_updater(
                 if composite_val in proteins_annot:
                     proteins_with_extra_annot += 1
                     extra_annot = proteins_annot[composite_val]
-                    output_extra.write(line.rstrip() + ";" + extra_annot + "\n")
-                    output_full.write(line.rstrip() + ";" + extra_annot + "\n")
+                    output_extra.write(line.rstrip() + ";" + extra_annot + pf_suffix + "\n")
```
Comment (KateSakharova) [inline bin/gff_mapping.py]: is it possible that `extra_annot` can be empty?

**Review by KateSakharova — APPROVED** [review]
Seems fine, I have some general thoughts:
1. Add new test csv-s into tests README (where you describe positive and negative csv-s)
2. I would restructure main README a bit. I would squeeze outputs more and do not show a big structure because you are describing it in the `outputs.md` page anyway. Just mention the most important outputs.
Do you have a link to Virify somewhere, btw?   
You have `Mobilome-only mode (v4 behaviour)` in usage.md. Would it be different for v5 which is coming?

File: bin/gff_mapping.py [hunk 2]
```diff
@@ -187,59 +189,127 @@ def mobilome_parser(mobilome_clean):
...
+    column re-ordering. Returns an empty dict if the file is missing, empty, or does
+    not contain the expected columns.
+    """
+    if is_file_empty(report_file):
+        logger.warning(f"Combined report is empty or missing: {report_file}")
+        return {}
+
+    summary_map = {}
+    with open_file(report_file) as handle:
+        header = handle.readline().rstrip("\n").split("\t")
```
Comment (mberacochea) [inline bin/gff_mapping.py:207]: Is the file a tsv? if you the `csv` module is better to parse

File: bin/gff_mapping.py [hunk 3]
```diff
@@ -346,7 +437,14 @@ def main():
         args.mobilome_gff
     )
 
-    # Adding the mobilome predictions to the user file
+    # Optional per-protein summary strings from the combined report
+    summary_map = parse_combined_report(args.combined_report) if args.combined_report else {}
+
+    # Optional renamed -> original contig mapping (Prodigal/tRNA baseline only)
+    names_equiv = mapping_names.names_map(args.contig_map)[0] if args.contig_map else {}
```
Comment (mberacochea) [inline bin/gff_mapping.py:450]: Why the `[0]`? Can this method just return one instead?

File: bin/gff_mapping.py [hunk 4]
```diff
@@ -187,59 +189,127 @@ def mobilome_parser(mobilome_clean):
...
+                protein_id = ""
+                for attr in l_line[8].split(";"):
+                    if attr.startswith("ID="):
+                        protein_id = attr[3:]
+                        break
+                pf_suffix = (
+                    f";pathofact2={summary_map[protein_id]}"
+                    if protein_id in summary_map
+                    else ""
+                )
```
Comment (mberacochea) [inline bin/gff_mapping.py:312]: In the future we should refactor this and use the toolkit. I fear that we may be missing errors or wrongly formatted bits in code like this.
The gff validator should catch them... but is a bit too late when that happens.
No action now, this is probably suitable for a hackathon

File: bin/pathofact2_report.py [hunk 5]
```diff
@@ -425,8 +430,22 @@ def build_rows(
         signalp = signalp_data.get(protein_id, "-")
         mge_type = protein_mge_map.get(protein_id, "-")
 
+        # Condensed per-gene tag string, tokens in fixed order: vf, arg, mge, bgc.
```
Comment (mberacochea) [inline bin/pathofact2_report.py:433]: But then the GFF doesn't have which BGC category it has (for example)?.

Comment (mberacochea) [reply bin/pathofact2_report.py:433]: Why do we need this? Not that I'm against it but the information is already there. This to me look like a downstream analysis bit.

File: docs/outputs.md [hunk 6]
```diff
@@ -101,9 +105,13 @@ GFF feature types and their Sequence Ontology mappings:
 
 **Seed proteins** — rows in the report — come from PathoFact2 and/or AMR GFF files. If both inputs are absent or empty, no report is generated. BGC annotations are only resolved for seed proteins; BGC-only or MGE-only proteins are not included as rows. MGE assignment requires ≥90% CDS overlap with a mobilome feature on the same contig. IPS input is optional; only SignalP entries are retained. Missing values are reported as `-`.
 
+Each protein's `summary_string` is also carried into the derived `*_mobilome_{clean,extra,full}.gff.gz` files as a `pathofact2=<summary_string>` attribute on the matching CDS.
+
 | Column | Description |
 |---|---|
 | `protein_id` | Protein identifier from Prodigal or the user-provided GFF |
+| `contig_id` | Contig the protein is located on |
+| `summary_string` | Condensed annotation tags for the gene, comma-joined in fixed order `vf,arg,mge,bgc`: `vf` (virulence — VFDB hit or PATHOFACT2 toxin/VF), `arg` (AMR), `mge` (within a mobile element), `bgc` (within a BGC). Every row carries at least `vf` or `arg` |
```
Comment (mberacochea) [inline docs/outputs.md:141]: Another option, which is easier to parse in some cases, is to add a few mode columns instead of having a summary string

File: tests/unit/test_gff_mapping.py [hunk 7]
```diff
@@ -213,10 +215,237 @@ def test_passenger_protein_detection(tmp_path):
...
+    it, the contigs would not match and no MGE/passenger would be written.
+    """
+    genes_gff = tmp_path / "genes.gff"
+    genes_gff.write_text(
+        "##gff-version 3\n"
+        "contig_1\tProdigal\tCDS\t1500\t1800\t.\t+\t0\tID=prot001\n"
+    )
+    # mobilome annotation is keyed by the ORIGINAL contig name
+    mobilome_annot = {
+        "NZ_real_1": ["NZ_real_1\tgeNomad\tvirus\t1000\t2000\t.\t+\t.\tID=virus001"]
```
Comment (mberacochea) [inline tests/unit/test_gff_mapping.py:508]: I find this strings hard to read, I think doing:
```python
"\t".join(
  "NZ_real_1",
  "geNomad",
  "virus",
  1000,
  2000,
  ".",
  "+",
  ".",
  "ID=virus001"
)
```
is easier to read.

File: workflows/mobilomeannotation.nf [hunk 8]
```diff
@@ -423,6 +410,36 @@ workflow MOBILOMEANNOTATION {
...
+        INTEGRATOR.out.mobilome_gff
+            .join(TRNAS_INTEGRATOR.out.merged_gff)
+            .join(RENAME.out.map_file)
+            .join(ch_user_proteins, remainder: true)
+            .map { row ->
+                // remainder:true appends a single null for the whole missing right side, so
+                // unmatched rows are [meta, mobilome, prodigal_gff, map_file, null] and matched
+                // rows are [meta, mobilome, prodigal_gff, map_file, user_gff, user_faa].
+                def (meta, mobilome_gff, prodigal_gff, map_file) = [row[0], row[1], row[2], row[3]]
+                def user_gff = row[4]
```
Comment (mberacochea) [inline workflows/mobilomeannotation.nf:429]: Can't you move this to the row instead?
```
.map { meta, mobilome_gff, prodigal_gff, map_file, user_gff ->
```
?

File: CHANGELOG.md [hunk 9]
```diff
@@ -3,6 +3,27 @@
...
+- The `summary_string` is now also carried into the derived GFFs as a
+  `pathofact2=<summary_string>` attribute on each matching CDS.
+- The derived `clean`/`extra`/`full` GFFs are now produced even when no user proteins are
+  provided, built from the Prodigal/tRNA genes GFF and named
+  `*_mobilome_{clean,extra,full}.gff.gz` (without the `_user` infix). With user proteins the
+  outputs keep their `*_user_mobilome_*` names.
+
+### Fixed
+
+- `*_user_mobilome_clean.gff` now starts with a `##gff-version 3` header and contains only
```
Comment (mberacochea) [inline CHANGELOG.md:22]: Is there a gff validation step?

File: README.md [hunk 10]
```diff
@@ -99,12 +100,26 @@ See [docs/usage.md](docs/usage.md) for skip flags, mobilome-only mode, and annot
 Results are written to `--outdir` (default: `results/`).
 
 ```
-{sample}/
-├── {sample}_combined_report.tsv
-├── {sample}_discarded_mge.txt
-├── {sample}_mobilome.fasta.gz
-├── {sample}_overlap_report.txt
+sample/
```
Comment (mberacochea) [inline README.md]: Why don't we just link this in the `outputs.md`, just to have less places to change things if we update the pipeline.

**Review by mberacochea — APPROVED** [review]
Thanks @Ales-ibt ! I left some questions in the PR, but nothing that should prevent the release I think.

File: README.md (hunk 10 above)
Comment (Ales-ibt) [reply README.md]: Agree, I have simplified this block

File: bin/pathofact2_report.py (hunk 5 above)
Comment (Ales-ibt) [reply bin/pathofact2_report.py:433]: The summary string is minimal. It is design to be used as a potential feature to search in the contig viewer, as we need the pathofact annotations to be searchable (this is a project requirement). The string is informative enough for the user to know in a glance if there are VF/ARG in the genome and if they are in the context of a BGC or an MGE without opening those tracks. Detailed information can be found in the pathofact report as you point this out.

File: docs/outputs.md (hunk 6 above)
Comment (Ales-ibt) [reply docs/outputs.md:141]: No need more columns, we have the detailed info in the summary. This is just to align the content of the report with what is found in the gff file

File: tests/unit/test_gff_mapping.py (hunk 7 above)
Comment (Ales-ibt) [reply tests/unit/test_gff_mapping.py:508]: Ah, this is a GFF entry. I will add a comment in the test

Comment (Ales-ibt) [reply tests/unit/test_gff_mapping.py:508]: Ah, you want to show it one line per row, ok, I will add that change

File: CHANGELOG.md (hunk 9 above)
Comment (Ales-ibt) [reply CHANGELOG.md:22]: We validate only the mobilome.gff because the rest of the outputs contains genes that, in the case of provided by the user, can contain strings in the attributes that violates the gff convention. That is the case for mettanotator and GCP outputs

File: bin/gff_mapping.py (hunk 3 above)
Comment (Ales-ibt) [reply bin/gff_mapping.py:450]: That function generates the [two-ways mapping dicts](https://github.com/EBI-Metagenomics/mobilome-annotation-pipeline/blob/70066330debdcc92a86dda51467729da5c050df3/bin/map_tools/mapping_names.py#L26). IMO opinion it is cleaner to generate both dicts in the same function than produce the inverted every time we need it

File: bin/gff_mapping.py (hunk 1 above)
Comment (Ales-ibt) [reply bin/gff_mapping.py]: No, `extra_annot` can't be empty beacuse it is read from `proteins_annot[composite_val]`, and entries are only added to `proteins_annot` (in `mobilome_parser`) when `len(extra_list) > 0` here https://github.com/EBI-Metagenomics/mobilome-annotation-pipeline/blob/70066330debdcc92a86dda51467729da5c050df3/bin/gff_mapping.py#L177

File: docs/outputs.md [hunk 11]
```diff
@@ -12,13 +12,17 @@ sample/
 ├── sample_overlap_report.txt
 ├── gff/
 │   ├── sample_mobilome.gff.gz
-│   ├── sample_user_mobilome_clean.gff.gz      # mobilome + matching CDSs
+│   │   # When user proteins+GFF are provided, the three derived files are named
+│   │   # sample_user_mobilome_{clean,extra,full}.gff.gz (baseline: the user GFF).
+│   │   # Otherwise they are named sample_mobilome_{clean,extra,full}.gff.gz
+│   │   # (baseline: the Prodigal/tRNA genes GFF). Each has matching .csi and .gzi indexes.
+│   ├── sample_user_mobilome_clean.gff.gz      # mobilome + genes covered by an MGE
```
Comment (Ge94) [inline docs/outputs.md]: Maybe I missed it - do you explain anywhere in the docs what clean/extra/full represent?

**Review by Ge94 — APPROVED** [review]
Yeeeeeah 🤘

File: docs/outputs.md (hunk 11 above)
Comment (Ales-ibt) [reply docs/outputs.md]: No, it is just mentioned in those comments. I will add a paragraph to explain better

File: workflows/mobilomeannotation.nf (hunk 8 above)
Comment (Ales-ibt) [reply workflows/mobilomeannotation.nf:429]: Yeah, we used to have a map there, we changed it because the .join(ch_user_proteins, remainder: true) emits variable-length tuples:
 - 6 elements when user proteins are present (…, user_gff, user_faa) and 
 - 5 in the remainder case (single trailing null). 

A fixed { meta, mobilome_gff, prodigal_gff, map_file, user_gff -> } signature did drop user_faa on matched rows and misbind across the two arities, so we use row -> + positional indexing here. This form was introduced in ebfb7579 when the no-user (Prodigal baseline) path was added; map_file followed in 51c7907a.

File: bin/gff_mapping.py (hunk 2 above)
Comment (Ales-ibt) [reply bin/gff_mapping.py:207]: Yes, this is a tsv file

**Comment by Ales-ibt** [comment]
Thank you for your comments! 

I had improved the tests/README and compacted the main README file by removing the results tree, as we are describing in detail in the docs.

I added the link to virify and other tools in the readme file.

Mobilome-only mode in version 5 (this one) would run exactly the same tools and versions that we run in version 4.

I am merging now so we can test this version on the chicken-gut catalogue 🥳
